src/software/backend/simulator_backend.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <mutex>
+
+#include "software/ai/world/world.h"
+#include "software/backend/backend.h"
+
+class SimulatorBackend : public Backend
+{
+   public:
+    static const std::string name;
+
+    /**
+     * Creates a new SimulatorBackend
+     *
+     * The SimulatorBackend will run a physics simulation to simulate and publish the
+     * World, and simulate the primitives given to it. The SimulatorBackend will only
+     * advance the simulation by a fixed amount each time primitives are received, rather
+     * than running at full-speed or in real-time. This is because we want the simulator
+     * to be deterministic and always provide the World at a "fixed" rate from the
+     * perspective of any consumers.
+     *
+     * @param simulation_time_step How far to step the simulation in time each time the
+     * simulation is advanced
+     * @param num_steps_per_primitive_update How many times to advance the simulation each
+     * time new primitives are received
+     */
+    explicit SimulatorBackend(
+        const Duration& simulation_time_step        = Duration::fromSeconds(1.0 / 60.0),
+        unsigned int num_steps_per_primitive_update = 2);
+
+    /**
+     * Returns the current state of the World in the simulated backend
+     *
+     * @return the current state of the World in the simulated backend
+     */
+    World getWorld();
+
+    /**
+     * Sets the state of the World in the simulated backend. All following simulation will
+     * follow from this newly provided world.
+     *
+     * @param new_world A new state of the World to set in the simulated backend
+     */
+    void setWorld(const World& new_world);
+
+   private:
+    /**
+     * Updates the simulation.
+     *
+     * This will advance the simulation by "simulation_time_step",
+     * "num_steps_per_primitive_update" times. We take several smaller steps if necessary
+     * to maintain the accuracy of the simulation. This will update the state of the World
+     * in the backend and send the updated values to all observers.
```

Comment:
What do you mean by accuracy here? ie. physics accuracy or robot/game state accuracy?

The physics accuracy should be taken care of by the physics engine as it . 

I think we would only need this in-case we want to "stop motion the ai" in case the physics engine can't keep up? Is that what you mean here?

Reply:
physics accuracy. I think the engine could take care of it, but it might affect things like collision behavior, and from my reading different engines may handle that better or worse than others.

See my comment above for the reasoning for not running everything real-time.

Reply:
Yea I think most physics engine will default to "sleep mode" where if theres no physics happening on the bodies, the bodies will "fall asleep" until waken up by a force, collision, etc.

But that makes sense since this is only the barebones skeleton implementation right now, we can talk more as we progress